src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs

```diff
@@ -0,0 +1,358 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.InteropServices;
+
+internal partial class Interop
+{
+    internal partial class mincore
+    {
+        internal const int COR_E_PLATFORMNOTSUPPORTED = unchecked((int)0x80131539);
+
+        // https://msdn.microsoft.com/en-us/library/windows/desktop/bb762188.aspx
+        [DllImport(Libraries.ShellFolders, CharSet = CharSet.Unicode, SetLastError = false, BestFitMapping = false, ExactSpelling = true, EntryPoint = "SHGetKnownFolderPath")]
+        private static extern int SHGetKnownFolderPath_ShellFolders(
+            [MarshalAs(UnmanagedType.LPStruct)] Guid rfid,
+            uint dwFlags,
+            IntPtr hToken,
+            out string ppszPath);
+
+        private static bool s_skipShellFolders;
+
+// Disabling the warning about availability.
+#pragma warning disable BCL0015
+        [DllImport(Libraries.Shell, CharSet = CharSet.Unicode, SetLastError = false, BestFitMapping = false, ExactSpelling = true, EntryPoint = "SHGetKnownFolderPath")]
+        private static extern int SHGetKnownFolderPath_Shell32(
+            [MarshalAs(UnmanagedType.LPStruct)] Guid rfid,
+            uint dwFlags,
+            IntPtr hToken,
+            out string ppszPath);
+#pragma warning restore BCL0015
+
+        internal static int SHGetKnownFolderPath(
+            [MarshalAs(UnmanagedType.LPStruct)] Guid rfid,
+            uint dwFlags,
+            IntPtr hToken,
+            out string ppszPath)
+        {
+            // The ShellFolders API set isn't available on Win7, while the direct Shell32 isn't available
+            // on some mincore versions of Windows. We'll attempt the newer API set version and fall back
+            // to the older "Shell32" if needed.
+            if (!s_skipShellFolders)
+            {
+                try
+                {
+                    return SHGetKnownFolderPath_ShellFolders(rfid, dwFlags, hToken, out ppszPath);
+                }
+                catch (TypeLoadException)
```

Comment:
> It does not work with the UWP AOT toolchain today (@yizhang82 - correct me if I am wrong).

Right. We only support early bound in AOT toolchain today, mostly because you can't LoadLibrary in UWP. This means any missing p/invoke would result in a OS load failure and crash the app immediately at launch. Having said that, it sounds like we should support late bound p/invokes in AOT to support this pattern. Unfortunately, this doesn't help you since we can't do late bound p/invokes anyway in UWP without Windows supporting LoadLibrary. And using LoadLibrary would mean failing WACK. Let's take this one offline and see what's the right action to take here.